File: smalivm/src/main/java/org/cf/smalivm/opcode/SGetOp.java
package org.cf.smalivm.opcode;

import org.cf.smalivm.StaticFieldAccessor;
import org.cf.smalivm.VirtualMachine;
import org.cf.smalivm.context.ExecutionContext;
import org.cf.smalivm.context.ExecutionNode;
import org.cf.smalivm.context.HeapItem;
import org.cf.smalivm.context.MethodState;
import org.cf.smalivm.type.VirtualField;
import org.jf.dexlib2.builder.MethodLocation;
import org.jf.dexlib2.iface.reference.FieldReference;
import org.jf.dexlib2.util.ReferenceUtil;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class SGetOp extends ExecutionContextOp {

    @SuppressWarnings("unused")
    private static final Logger log = LoggerFactory.getLogger(SGetOp.class.getSimpleName());

    private final int destRegister;
    private final FieldReference fieldReference;
    private final VirtualField actualField;
    private final VirtualMachine vm;

    SGetOp(MethodLocation location, MethodLocation child, int destRegister, FieldReference fieldReference,
           VirtualField actualField, VirtualMachine vm) {
        super(location, child);

        this.destRegister = destRegister;
        this.fieldReference = fieldReference;
        this.actualField = actualField;
        this.vm = vm;
    }

    @Override
    public void execute(ExecutionNode node, ExecutionContext context) {
        StaticFieldAccessor accessor = vm.getStaticFieldAccessor();
        HeapItem item = accessor.getField(context, actualField);
        MethodState mState = context.getMethodState();
        mState.assignRegister(destRegister, item);
    }

    @Override
    public String toString() {
        return getName() + " r" + destRegister + ", " + ReferenceUtil.getFieldDescriptor(fieldReference);
    }

}

File: smalivm/src/main/java/org/cf/smalivm/opcode/SGetOp.kt
package org.cf.smalivm.opcode

import org.cf.smalivm.configuration.Configuration
import org.cf.smalivm.dex.SmaliClassLoader
import org.cf.smalivm.type.ClassManager
import org.cf.smalivm.type.VirtualField
import org.cf.smalivm2.ExecutionNode
import org.cf.smalivm2.ExecutionState
import org.cf.smalivm2.UnresolvedChild
import org.jf.dexlib2.builder.MethodLocation
import org.jf.dexlib2.iface.instruction.formats.Instruction21c
import org.jf.dexlib2.iface.reference.FieldReference
import org.slf4j.LoggerFactory

class SGetOp internal constructor(
    location: MethodLocation,
    private val destRegister: Int,
    private val field: VirtualField
) : Op(location) {

    override val registersReadCount = 1
    override val registersAssignedCount = 1

    override fun execute(node: ExecutionNode): Array<out UnresolvedChild> {
        return if (!node.state.isClassInitialized(field.definingClass)) {
            staticInitClass(field.definingClass, node.classManager, node.classLoader, node.configuration)
        } else {
            val item = node.state.peekField(field)
            node.state.assignRegister(destRegister, item)
            finish()
        }
    }

    override fun toString() = "$name r$destRegister, $field"

    companion object : OpFactory {
        private val log = LoggerFactory.getLogger(SGetOp::class.java.simpleName)

        override fun build(
            location: MethodLocation,
            classManager: ClassManager,
            classLoader: SmaliClassLoader,
            configuration: Configuration
        ): Op {
            val instr = location.instruction as Instruction21c
            val destRegister = instr.registerA
            val fieldReference = instr.reference as FieldReference
            val fieldClass = classManager.getVirtualClass(fieldReference.definingClass)
            val field = fieldClass.getField(fieldReference.name)!!
            return SGetOp(location, destRegister, field)
        }
    }
}
